v2 blog post (#351)
vercel/swr-site: pages/blog/swr-v2.en-US.mdx

---
image:
description: 'We are excited to announce SWR v2.0 release, with various improvements of DX and new features like preloading, stale, optimistic update, etc.'
date:
---

import Callout from 'nextra-theme-docs/callout'
import Bleed from 'nextra-theme-docs/bleed'

import Authors, { Author } from 'components/authors'
import Video from 'components/video'

# Announcing SWR 2.0

<Authors date="">
</Authors>

We are excited to introduce SWR v2.0, including features to improve the use experience of your application.

## What’s New

> ### Better React 18 Support

Reviewer: We can move this section to the bottom, since there's no new API change, basically improvement. We want to show users the explict changes first which can make it a bit more attractive

  if (isLoading) return <div>loading...</div>
  return <div>hello {data.name}!</div>
}
```

<Callout emoji="📝">
  We have added the new [Understanding SWR](/docs/advanced/understanding) page to describe how SWR returns values, which includes the difference between `isValidating` and `isLoading`, and how to combine them to improve user exeprience.
</Callout>

### Lazy Updating

`keepPreviousData` is a new option for keeping the previous fetched data. This improves the UX a lot when you fetch data based on continuous user actions, e.g. real-time search when typing. More information can be found [here](/docs/advanced/understanding#return-previous-data-for-better-ux).

```jsx
function Search() {
  const [search, setSearch] = React.useState('');

  const { data, isLoading } = useSWR(`/search?q=${search}`, fetcher, {
    keepPreviousData: true
  });

  return (
    <div>
      <input
        type="text"
        value={search}
        onChange={(e) => setSearch(e.target.value)}
        placeholder="Search..."
      />

      <div className={isLoading ? "loading" : ""}>
        {data?.products.map(item => <Product key={item.id} name={item.name} />)
      </div>
    </div>
  );
}
```

> 

Reviewer: it's bit hard to find out it's a video on the website, since the cover picture is almost same color `#fff` with light mode background. adding a title to make it more explicit
```suggestion
#### Video Demo 
```

---
image:

> description: 'We are excited to introduce SWR v2.0, including features to improve the use experience of your application.'

Reviewer: ```suggestion
description: 'We are excited to announce SWR v2.0 release, with various improvements of DX and new features like preloading, stale, optimistic update, etc.'
```

date:
---

import Callout from 'nextra-theme-docs/callout'
import Bleed from 'nextra-theme-docs/bleed'

import Authors, { Author } from 'components/authors'
import Video from 'components/video'

# Announcing SWR 2.0

<Authors date="">
</Authors>

We are excited to introduce SWR v2.0, including features to improve the use experience of your application.

## What’s New

### Better React 18 Support

> SWR now uses `useSyncExternalStore` and `startTransition` internally to improve React 18 support, which ensures stronger UI consistency in concurrent rendering with React 18. The change is an internal implementation detail, so developers using SWR do not have to care about the change.

Reviewer: ```suggestion
SWR now uses `useSyncExternalStore` and `startTransition` internally to improve the performance and manage states while integrating with React 18, which ensures stronger UI consistency under concurrent rendering mode. This change doesn't require any implicity updates on the interface perspective. Developers will benefit from the updates directly.
```

> ### isLoading State

Reviewer: ```suggestion
### New Loading State Indicator
```

> `isLoading` is a new return value that indicates if there's an ongoing request and no "loaded data". Previously, you have to check `data` and `error` are `undefined`, but now you can use the `isLoading` value.

Reviewer: Since previously users can also use `isValidating`, we might want to clarify what's the difference.

Author: @huozhi Thank you! I'll create a new section for the difference between `isLoading` and `isValidating` on the Understanding SWR page and put the link in the blog post.

> You can improve the user experience by combining with `isLoading` and `isValidating` like this.

Reviewer: I thinnk we can only have a simple example here, and link to the page with details of how `isLoading` and `isValidating` work.

Author: @huozhi Thank you for your feedback! I've moved the details to the Understanding SWR page and simplified the isLoading example.

```jsx
function Stock() {
  const { data, isLoading, isValidating } = useSWR(STOCK_API, fetcher, {
    refreshInterval: 3000
  });

  // If it's still loading the initial data, there is nothing to display.
  // We return a skeleton here.
  if (isLoading) return <div className="skeleton" />;

  // Otherwise, display the data and a spinner that indicates a background
  // revalidation.
  return (
    <>

>       <div>AAPL ${data}</div>

Reviewer: not sure what's AAPL, is it apple stock? wonder we can have less obscure working here to make the example more straightforward

Author: 👍 I've picked up the example from https://github.com/vercel/swr/releases/tag/2.0.0-beta.1

Reviewer: yeah but if other languages want to translate, less special abbr words will make it easy. Maybe just `<div>${data}</div>`

      {isValidating ? <div className="spinner" /> : null}
    </>
  );
}
```

![An example using the isLoading state](https://user-images.githubusercontent.com/3676859/163695538-9ed6232d-808c-4c83-8606-0dd22bce508d.gif)

You can find the full code for this example here: https://codesandbox.io/s/swr-isloading-v8dfpy.

<Callout emoji="📝">
  We have added the new [Understanding SWR](/docs/advanced/understanding) page to describe how SWR returns values.
</Callout>

> ### keepPreviousData Option

Reviewer: ```suggestion
### Lazy Updating
```

```jsx
import { mutate } from 'swr'
// Or from the hook if you customized the cache provider:
// { mutate } = useSWRConfig()

mutate(
  key => typeof key === 'string' && key.startsWith('/api/item?id='),
  undefined,
  { revalidate: true }
)
```

### Preload API

`preload` is a new API for prefetching data programmatically. Prefetching data is important to improve user experiences and avoid waterfall problems. You can call `preload` before rendering. More information can be found [here](/docs/prefetching).

```jsx
import useSWR, { preload } from 'swr'

const fetcher = (url) => fetch(url).then((res) => res.json())
// You can call the preload function in anywhere
preload('/api/user', fetcher)
```

### Functional optimisticData

`optimisticData` is an option of mutation, which is useful to implement optimistic UI. In v2, you can pass a function as the `optimisticData` option, which accepts current data and returns the new client cache data. More information can be found [here](/docs/mutation#optimistic-updates).

```jsx
mutate('/api/user', updateUserName(newName), {
  optimisticData: user => ({ ...data, name: newName }),
  rollbackOnError: true
});
```

> ### Function as SWRConfig value

Reviewer: ```suggestion
### Functional SWR Config
```

> The value prop of `SWRConfig` accepts a function that receives the parent config and returns a new config. More information can be found [here](/docs/global-configuration).

Reviewer: ```suggestion
`SWRConfig` can now accept a function for `value` prop that receives the parent config and returns a new config. This empowers user flexibility to configure the SWR config especially in nested `<SWRConfig>` cases. More information can be found [here](/docs/global-configuration).
```

```jsx
<SWRConfig
  value={parent => ({
    dedupingInterval: parent.dedupingInterval * 5,
    refreshInterval: 100,
  })}
>
  <Page />
</SWRConfig>
```

### DevTools

> SWRDevTools is a browser extension for SWR, which helps to debug your SWR cache and the results of your fetcher. More information can be found [here](/docs/advanced/devtools).

Reviewer: ```suggestion
SWRDevTools is a browser extension for SWR, which helps to debug your SWR cache and the results of your fetcher. Checkout [devtools](/docs/advanced/devtools) section for how to use devtools in your application.
```

## Migration Guide

### Fetcher no longer accepts multiple arguments

`key` is now passed as a single argument.

```diff
- useSWR([1, 2, 3], (a, b, c) => {
+ useSWR([1, 2, 3], ([a, b, c]) => {
  assert(a === 1)
  assert(b === 2)
  assert(c === 3)
})
```

> ### Mutate accepts a filter function

Reviewer: ```suggestion
### Multiple Mutations
```

`mutate` now accepts [a filter function](/blog/swr-v2#mutate-multiple-keys). Previously, the function behaved as returing a key.

```diff
- mutate(() => '/api/item') // a function to return a key
+ mutate(key => typeof key === 'string' && key.startsWith('/api/item?id=')) // a filter function
```

> ### The Cache interface requires a `keys()`

Reviewer: ```suggestion
### New Required Property `keys()` for Cache Interface
```

When you use your own cache implementation, the Cache interface now requires a `keys()` method that returns all keys in the cache object, similar to the JavaScript Map instances.

```diff
interface Cache<Data> {
  get(key: string): Data | undefined
  set(key: string, value: Data): void
  delete(key: string): void
+ keys(): IterableIterator<string>
}
```

### Internal Structure Of The Cached Data

The internal structure of the cache data will be an object that holds all the current states.

```diff
- assert(cache.get(key) === data)
+ assert(cache.get(key) === { data, error, isValidating })

// getter
- cache.get(key)
+ cache.get(key)?.data

// setter
- cache.set(key, data)
+ cache.set(key, { ...cache.get(key), data })
```

<Callout emoji="🚨" type="error">
  You should not write to the cache directly, it might cause undefined behavior.
</Callout>

### SWRConfig.default → SWRConfig.defaultValue

> `SWRConfig.defaultValue` is the property for accessing the default value.

Reviewer: ```suggestion
`SWRConfig.defaultValue` is the property for accessing the default SWR config.
```

```diff
- SWRConfig.default
+ SWRConfig.defaultValue
```

### Type InfiniteFetcher is renamed to SWRInfiniteFetcher

```diff

> - import { InfiniteFetcher } from 'swr/infinite'
> + import { SWRInfiniteFetcher } from 'swr/infinite'

Reviewer: ```suggestion
- import type { InfiniteFetcher } from 'swr/infinite'
+ import type { SWRInfiniteFetcher } from 'swr/infinite'
```
